### PR TITLE
Allow this.state assignment on constructor

### PR DESCRIPTION
**Pull request by burabure**
see #832

### CODE REVIEW

File: lib/rules/no-direct-mutation-state.js [hunk 1]
```diff
@@ -64,26 +71,32 @@ module.exports = {
...
+          inConstructor = false;
         }
+      },
+
+      'Program:exit': function () {
+        const list = components.list();
+
+        Object.keys(list).forEach((key) => {
+          if (isValid(list[key])) {
+            return;
```
Comment (ljharb) [inline lib/rules/no-direct-mutation-state.js]: Might be simpler to invert the if and put the report inside

File: docs/rules/no-direct-mutation-state.md [hunk 2]
```diff
@@ -1,8 +1,10 @@
 # Prevent direct mutation of this.state (react/no-direct-mutation-state)
 
-NEVER mutate `this.state` directly, as calling `setState()` afterwards may replace
+Don't mutate `this.state` directly, as calling `setState()` afterwards may replace
 the mutation you made. Treat `this.state` as if it were immutable.
 
+The only place that it's acceptable to assign this.state is on a ES6 class component constructor
```
Comment (ljharb) [inline docs/rules/no-direct-mutation-state.md]: s/on/in, and add a period at the end. Also lets backtick `class`

**Comment by burabure** [comment]
@ljharb ready for review

File: lib/rules/no-direct-mutation-state.js [hunk 3]
```diff
@@ -49,11 +49,18 @@ module.exports = {
     // --------------------------------------------------------------------------
     // Public
     // --------------------------------------------------------------------------
+    var inConstructor = false;
```
Comment (ljharb) [inline lib/rules/no-direct-mutation-state.js]: let's use `let` here.

**Comment by burabure** [comment]
@ljharb ready for review!

**Comment by valentin-radulescu-hs** [comment]
👍

**Comment by supra28** [comment]
Has this been released yet?

**Comment by ljharb** [comment]
No. You can tell by clicking on the commit hash - 61b65a0 - and seeing that it's in master, but there's no version tags that it's in.

**Comment by supra28** [comment]
@ljharb  Thanks

**Comment by jsg2021** [comment]
I was under the impression `this.state = {}` was a valid state reset/replacement.

**Comment by ljharb** [comment]
Nope; React would have no way to know you'd done that, so it wouldn't trigger a rerender. You should only change state with `this.setState`.

**Comment by supra28** [comment]
@ljharb I have updated the plugin to 7.3.0 and I still get this error :(

**Comment by ljharb** [comment]
@supra28 i'm not sure what error you mean. can you file a new issue?

**Pull request merged**